### ISSUE TITLE
Train Time Increase with gradient accumulation for pytorch_lighting with version release/1.3.x/

### ISSUE THREAD

**Issue by sIncerass** [PyTorchLightning/pytorch-lightning#8066]
### To Reproduce

[**BoringModel**](https://colab.research.google.com/drive/1h6TH5JZ0PdhlfOlFFdoEaqWzvs5kWoJa?usp=sharing) also reproduces the same bug. 

```
import os
import torch
from torch import nn
import torch.nn.functional as F
from torchvision.datasets import MNIST
from torch.utils.data import DataLoader, random_split
from torchvision import transforms
import pytorch_lightning as pl

class LitAutoEncoder(pl.LightningModule):

    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(nn.Linear(28 * 28, 128), nn.ReLU(), nn.Linear(128, 3))
        self.decoder = nn.Sequential(nn.Linear(3, 128), nn.ReLU(), nn.Linear(128, 28 * 28))

    def forward(self, x):
        # in lightning, forward defines the prediction/inference actions
        embedding = self.encoder(x)
        return embedding

    def training_step(self, batch, batch_idx):
        # training_step defines the train loop. It is independent of forward
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = F.mse_loss(x_hat, x)
        self.log('train_loss', loss)
        self.trainer.profiler.describe()
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer

dataset = MNIST(os.getcwd(), download=True, transform=transforms.ToTensor())
train, val = random_split(dataset, [55000, 5000])

autoencoder = LitAutoEncoder()
trainer = pl.Trainer(profiler="simple", accumulate_grad_batches=4, gpus=1, log_every_n_steps=50, distributed_backend='ddp',)
trainer.profiler.dirpath = "/net/s3/allennlp/shengs/doubling/scripts/ptl_test"
trainer.profiler.filename = 'profiler_c1'
trainer.fit(autoencoder, DataLoader(train), DataLoader(val))
```

### Expected behavior
The time in "model forward" will be increased when we set the profiler and gradient accumulation more than 1 using the boarding model or the simple model I provided above. 

### Environment

 - PyTorch Version (1.8,1):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version: 3.8.10
 - CUDA/cuDNN version: 11.2
 - GPU models and configuration: Quadro RTX 8000
 - Any other relevant information:

**Comment by awaelchli** [comment]
which versions of lightning did you compare?

**Comment by ibeltagy** [comment]
I have seen this issue with version 1.0.8. @sIncerass had the same issue with version 1.3.5. In both cases, the time per step gets longer as the model trains longer. We tried the profiler and it looks like the forward pass is what is slowing down.

**Comment by awaelchli** [comment]
Thanks for the clarification, that's an important detail you left out there in the post. So as you say there is no difference in the versions. **The comparision is the training time from epoch to epoch.**

Have you also tried with the master branch? There was an internal logging re-design which also is supposed to bring perf improvements.

**Comment by ibeltagy** [comment]
>  The comparision is the training time from epoch to epoch.

Not even epochs. Every step is a tiny bit slower than the step before it. Will try to send you a screenshot for the tensorboard log. 

> Have you also tried with the master branch? There was an internal logging re-design which also is supposed to bring perf improvements.

No we didn't. Is it a general performance improvement or something specific to gradient accumulation and the gradual slowing down?

**Comment by ibeltagy** [comment]
Here are my logs from version 1.0.8
![image](https://user-images.githubusercontent.com/2287797/122948909-1c341b80-d330-11eb-8e68-4ce7ba3e104a.png)
and here with smoothing 
![image](https://user-images.githubusercontent.com/2287797/122949003-2e15be80-d330-11eb-95aa-bf0bfdaaf3a9.png)
At the beginning, it is 0.87s/batch. At step 25k, it is 1.38s/batch. 

@sIncerass has better logs using the recently added profiler, but it is pretty much the same story.

**Comment by awaelchli** [comment]
I took your repro script (btw no access to the link) and modified it. On CPU, can't reproduce your observations on master. I get `model_forward` times around 0.0004 seconds across all these settings I tried:

```python
settings = [
    # accumulation, max_steps
    (1, 1000),
    (4, 250),
    (8, 125),
    (8, 250),
    (8, 500),
    (8, 1000),
]
```

Here is the full script. I will now run the same on GPU and check also 1.3.x. 
```python

import torch
from torch import nn
import torch.nn.functional as F
from torchvision.datasets import MNIST
from torch.utils.data import DataLoader, random_split
from torchvision import transforms
import pytorch_lightning as pl


class LitAutoEncoder(pl.LightningModule):

    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(nn.Linear(28 * 28, 128), nn.ReLU(), nn.Linear(128, 3))
        self.decoder = nn.Sequential(nn.Linear(3, 128), nn.ReLU(), nn.Linear(128, 28 * 28))

    def forward(self, x):
        embedding = self.encoder(x)
        return embedding

    def training_step(self, batch, batch_idx):
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = F.mse_loss(x_hat, x)
        self.log('train_loss', loss)
        # self.trainer.profiler.describe()  # why was this here??
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


dataset = MNIST("./MNIST", download=True, transform=transforms.ToTensor())
train, val = random_split(dataset, [55000, 5000])

autoencoder = LitAutoEncoder()


settings = [
    # accumulation, max_steps
    (1, 1000),
    (4, 250),
    (8, 125),
    (8, 250),
    (8, 500),
    (8, 1000),
]

for accumulation, max_steps in settings:
    trainer = pl.Trainer(
        profiler="simple",
        accumulate_grad_batches=accumulation,
        log_every_n_steps=50,
        max_steps=max_steps,
    )
    trainer.profiler.dirpath = "lightning_logs"
    trainer.profiler.filename = f'fprofiler_accumulation_{accumulation}_{max_steps}'
    trainer.fit(autoencoder, DataLoader(train), DataLoader(val))

```

**Comment by ibeltagy** [comment]
thanks, @awaelchli. I think you need to run it a lot longer than a few hundred steps. From the tensorboard log earlier, you need like 10k steps for the slowdown to become noticeable. 
Just run the same 1GPU, 4 accumulation steps, v.1.3.5 configuration but for a lot longer. You need to log to tensorboard because the difference can go unnoticed in the progress bar.

**Comment by ibeltagy** [comment]
> EDIT: my runs are probably too short to see a noticable differnece, gotta reproduce runs like the ones from #8066 (comment)

Just saw your edit. Agree

**Comment by sIncerass** [comment]
Yeah, more examples with gradient accumulation where the model_forward time increase are added here.
![Screen Shot 2021-06-22 at 2 42 30 PM](https://user-images.githubusercontent.com/22538462/123003481-1c500d80-d368-11eb-937b-8c266174ec39.png)
 ![Screen Shot 2021-06-22 at 2 42 03 PM](https://user-images.githubusercontent.com/22538462/123003435-0b9f9780-d368-11eb-9ba3-51207ea47e02.png)
![Screen Shot 2021-06-22 at 2 43 34 PM](https://user-images.githubusercontent.com/22538462/123003588-3f7abd00-d368-11eb-8551-374a3ef0ddd5.png)

**Comment by awaelchli** [comment]
After 100k iterations I'm still seeing flat curves. Are you aware that you have a `self.trainer.profiler.describe()` call in `training_step()`? Can you confirm and promise 100% that you have shared the code that you have run to produce the logs you are showing, and that you have stripped away only the logging code that you produced the screenshots with?

**Comment by sIncerass** [comment]
Sorry @awaelchli, can you elaborate on the GPU machine and the scripts? Yes, we have the `self.trainer.profiler.describe()` in the `training_step()`.

**Comment by awaelchli** [comment]
I'm not so much familiar with the profilers and how they are supposed to work but you definitely need to remove
`self.trainer.profiler.describe()` from the training step, this was slowing down my iterations so I removed it as I thought you added this for debugging. 

My script was derived from yours directly, but since you didn't add the visualization and timing code I added that myself:
```python
import os
import time

import torch
from torch import nn
import torch.nn.functional as F
from torchvision.datasets import MNIST
from torch.utils.data import DataLoader, random_split
from torchvision import transforms
import pytorch_lightning as pl
from pytorch_lightning.loggers import WandbLogger


class LitAutoEncoder(pl.LightningModule):

    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(nn.Linear(28 * 28, 128), nn.ReLU(), nn.Linear(128, 3))
        self.decoder = nn.Sequential(nn.Linear(3, 128), nn.ReLU(), nn.Linear(128, 28 * 28))

    def forward(self, x):
        # in lightning, forward defines the prediction/inference actions
        embedding = self.encoder(x)
        return embedding

    def training_step(self, batch, batch_idx):
        # training_step defines the train loop. It is independent of forward
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = F.mse_loss(x_hat, x)
        self.log('train_loss', loss)
        # self.trainer.profiler.describe()  # remove this, why is it here?
        return loss

    def on_train_batch_start(self, *args, **kwargs):
        self._start = time.monotonic()

    def on_train_batch_end(self, *args, **kwargs):
        delta = time.monotonic() - self._start
        self.log("time", delta, on_step=True, on_epoch=False)

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


dataset = MNIST(os.getcwd(), download=True, transform=transforms.ToTensor())
train, val = random_split(dataset, [55000, 5000])

logger = WandbLogger(project="accumulation-perf", name="orig")

autoencoder = LitAutoEncoder()
trainer = pl.Trainer(logger=logger, profiler="simple", accumulate_grad_batches=4, gpus=1, log_every_n_steps=50, distributed_backend='ddp',)
trainer.profiler.dirpath = "lightning_logs"
trainer.profiler.filename = f'fprofiler_accumulation_orig'
trainer.fit(autoencoder, DataLoader(train), DataLoader(val))
```

Here is the preliminary report:
https://wandb.ai/awaelchli/accumulation-perf/reports/Gradient-Accumulation-Performance-Debugging--Vmlldzo3OTg3NDE?accessToken=9rnvzcpallhgl0khua76lo5x1g365z1pxjzk0w0o6og912yqtejui9bfzxa2iwe5

**Comment by sIncerass** [comment]
I use the

**Comment by ibeltagy** [comment]
Thanks, @awaelchli for your help. Removing the profiler and moving the timing to `on_train_batch_start `, `on_train_batch_end` fixed the problem with v.1.3.5. We don't know what was the problem with the older, v1.0.8 version but we don't care, we are not using it anymore. 
It is an interesting coincidence that the issue manifests in v1.0.8 and in v1.3.5+profiler but not in v1.3.5. Any guesses why this might be the case?

**Comment by awaelchli** [comment]
hmm interesting, there have been changes to the profilers but I haven't kept up with those. Not sure what was going on here. 
the resolution is not very satisfactory here, because I haven't seen this increasing overhead due to the profiler. I shall go back to 1.3.5 and test again
